packages/core/src/model.ts

```diff
@@ -6,24 +6,82 @@ import type {
   ExtractEvent,
   EventObject
 } from './types';
+import { mapValues } from './utils';
 
-export interface ContextModel<TContext, TEvent extends EventObject> {
+export interface ContextModel<
+  TContext,
+  TEvent extends EventObject,
+  TEM extends EventCreatorMap<TEvent> = EventCreatorMap<TEvent>
+> {
   initialContext: TContext;
   assign: <TEventType extends TEvent['type'] = TEvent['type']>(
     assigner:
       | Assigner<TContext, ExtractEvent<TEvent, TEventType>>
       | PropertyAssigner<TContext, ExtractEvent<TEvent, TEventType>>,
     eventType?: TEventType
   ) => AssignAction<TContext, ExtractEvent<TEvent, TEventType>>;
+  events: FullEventCreatorMap<TEM>;
+  types: {
+    context: TContext;
+    events: TEvent;
+  };
   reset: () => AssignAction<TContext, any>;
 }
 
+type EventCreatorMap<TEvent extends EventObject> = {
+  [key in TEvent['type']]: (
+    ...args: any[]
+  ) => Omit<TEvent & { type: key }, 'type'>;
+};
+
+type FullEventCreatorMap<
+  TEM extends EventCreatorMap<any>,
+  TE extends EventObject = GetEvents<TEM>
+> = TEM extends any
+  ? {
+      [K in keyof TEM]: (
+        ...args: Parameters<TEM[K]>
+      ) => TE extends { type: K } ? TE : never;
+    }
+  : never;
+
+type Expand<T> = T extends infer O ? { [K in keyof O]: O[K] } : never;
+type Distribute<U, V extends Record<string, () => object>> = U extends keyof V &
+  string
+  ? Expand<{ type: U } & ReturnType<V[U]>>
+  : never;
+
+type GetEvents<T extends EventCreatorMap<any>> = Distribute<keyof T, T>;
+
 export function createModel<TContext, TEvent extends EventObject>(
```

Comment:
q: what is the intention here - is the model supposed to restrict accepted `TEvent` for assigners when `eventCreators` are used? previously there were no such restrictions (as obviously `eventCreators` didn't exist) so assigners could be used with "arbitrary" `TEvent` that could just be configured through the generic. So this part is being "lost" - which isn't necessarily a bad thing though.

Reply:
The intention is to not provide `TEvent` (or any generics) at all when specifying event creators.